README.md
[![Bonsai Asset Badge](https://img.shields.io/badge/Sensu%20Teams%20Handler-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/mildgrim/sensu-teams-handler) [![Build Status](https://travis-ci.org/mildgrim/sensu-teams-handler.svg?branch=master)](https://travis-ci.org/mildgrim/sensu-teams-handler)

# Sensu Teams Handler

- [Overview](#overview)
- [Usage examples](#usage-examples)
  - [Help output](#help-output)
  - [Environment variables](#environment-variables)
  - [Templates](#templates)
  - [Annotations](#annotations)
- [Configuration](#configuration)
  - [Asset registration](#asset-registration)
  - [Handler definition](#handler-definition)
  - [Check definition](#check-definition)
- [Installation from source and contributing](#installation-from-source-and-contributing)

## Overview


The [Sensu Teams Handler][0] is a [Sensu Event Handler][3] that sends event data
to a configured Teams channel.

## Usage examples

### Help output

Help:

```
Usage:
  sensu-teams-handler [flags]
  sensu-teams-handler [command]

Available Commands:
  help        Help about any command
  version     Print the version number of this plugin

Flags:
  -s, --sender string                 The name that messages will be sent as (default "Sensu")
  -d, --description-template string   The Teams notification output template, in Golang text/template format (default "{{ .Check.Output }}")
  -h, --help                          help for sensu-teams-handler
  -u, --sensu-url string              A URL to the Sensu dashboard (default "http://localhost:3000")
  -t, --is-test string                Specify if this is a test run (default "false")
  -w, --webhook-url string            The webhook url to send messages to
```

### Environment variables

|Argument               |Environment Variable       |
|-----------------------|---------------------------|
|--webhook-url          |TEAMS_WEBHOOK_URL          |
|--sender               |TEAMS_SENDER               |
|--is-test              |TEAMS_IS_TEST              |
|--sensu-url            |TEAMS_SENSU_URL            |
|--description-template |TEAMS_DESCRIPTION_TEMPLATE |


**Security Note:** Care should be taken to not expose the webhook URL for this handler by specifying it
on the command line or by directly setting the environment variable in the handler definition.  It is
suggested to make use of [secrets management][7] to surface it as an environment variable.  The
handler definition above references it as a secret.  Below is an example secrets definition that make
use of the built-in [env secrets provider][8].

```yml
---
type: Secret
api_version: secrets/v1
metadata:
  name: teams-webhook-url
spec:
  provider: env
  id: TEAMS_WEBHOOK_URL
```

### Templates

This handler provides options for using templates to populate the values
provided by the event in the message sent via Teams. More information on
template syntax and format can be found in [the documentation][9]


### Annotations

All arguments for this handler are tunable on a per entity or check basis based
on annotations. The annotations keyspace for this handler is
`sensu.io/plugins/teams/config`.

**NOTE**: Due to [check token substituion][10], supplying a template value such
as for `description-template` as a check annotation requires that you place the
desired template as a [golang string literal][11] (enlcosed in backticks)
within another template definition.  This does not apply to entity annotations.

#### Examples

To customize the sender for a given entity, you could use the following
sensu-agent configuration snippet:

```yml
# /etc/sensu/agent.yml example
annotations:
  sensu.io/plugins/teams/config/sender: 'server1'
```

## Configuration

### Asset registration

Assets are the best way to make use of this handler. If you're not using an asset, please consider doing so! If you're using sensuctl 5.13 or later, you can use the following command to add the asset:

`sensuctl asset add mildgrim/sensu-teams-handler`

If you're using an earlier version of sensuctl, you can download the asset
definition from [this project's Bonsai Asset Index
page][6].

### Handler definition

Create the handler using the following handler definition:

```yml
---
api_version: core/v2
type: Handler
metadata:
  namespace: default
  name: teams
spec:
  type: pipe
  command: sensu-teams-handler --sensuUrl 'http://localhost:3000' --sender 'Sensu'
  filters:
  - is_incident
  runtime_assets:
  - sensu/sensu-teams-handler
  secrets:
  - name: TEAMS_WEBHOOK_URL
    secret: teams-webhook-url
  timeout: 10
```
**Note**: The library used in the Sensu SDK for this plugin requires that if your Teams webhook URL is listed as an environment variable, the URL cannot be surrounded by quotes. 

**Security Note**: The Teams webhook URL should always be treated as a security
sensitive configuration option and in this example, it is loaded into the
handler configuration as an environment variable using a [secret][5]. Command
arguments are commonly readable from the process table by other unprivaledged
users on a system (ex: ps and top commands), so it's a better practise to read
in sensitive information via environment variables or configuration files on
disk. The --webhook-url flag is provided as an override for testing purposes.

### Check definition

```
api_version: core/v2
type: CheckConfig
metadata:
  namespace: default
  name: dummy-app-healthz
spec:
  command: check-http -u http://localhost:8080/healthz
  subscriptions:
  - dummy
  publish: true
  interval: 10
  handlers:
  - teams
```

### Proxy Support

This handler supports the use of the environment variables HTTP_PROXY,
HTTPS_PROXY, and NO_PROXY (or the lowercase versions thereof). HTTPS_PROXY takes
precedence over HTTP_PROXY for https requests.  The environment values may be
either a complete URL or a "host[:port]", in which case the "http" scheme is assumed.

## Installing from source and contributing

Download the latest version of the sensu-teams-handler from [releases][4],
or create an executable from this source.

### Compiling

From the local path of the sensu-teams-handler repository:
```
go build
```

To contribute to this plugin, see [CONTRIBUTING](https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md)

[0]: https://github.com/mildgrim/sensu-teams-handler
[1]: https://github.com/sensu/sensu-go
[3]: https://docs.sensu.io/sensu-go/latest/reference/handlers/#how-do-sensu-handlers-work
[4]: https://github.com/sensu/sensu-teams-handler/releases
[5]: https://docs.sensu.io/sensu-go/latest/reference/secrets/
[6]: https://bonsai.sensu.io/assets/mildgrim/sensu-teams-handler
[7]: https://docs.sensu.io/sensu-go/latest/guides/secrets-management/
[8]: https://docs.sensu.io/sensu-go/latest/guides/secrets-management/#use-env-for-secrets-management
[9]: https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-process/handler-templates/
[10]: https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-schedule/checks/#check-token-substitution
[11]: https://golang.org/ref/spec#String_literals


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed
- Refactoring from a Slack handler to a Microsoft Teams handler
- Added dependency to https://github.com/atc0005/go-teams-notify

## [1.5.0] - 2021-03-022

### Changed
- Fixed README
- Update to use plugin sdk v0.12: adds support for hostname function for use in template

## [1.4.0] - 2021-02-08

### Changed
- Update to use plugin sdk 0.11.0
- Mark --webhook-url argument as secret, so if set via SLACK_WEBHOOK_URL, value will not leak into usage message.
- Q1 '21 handler maintenance:
  - Updated modules (go get -u && go mod tidy)
  - Updated GitHub Actions: Added Lint action
  - Updated build to Go 1.14
  - Fix linter error by having template error go to stdout
  - Change return from sendMessage to capture and embellish any error
  - Output log message that message was sent to Slack
  - Fix TestSendMessage to include template
  - README updates

## [1.3.2] - 2020-05-18

### Changed
- Support for newline (\n) expansion in description template
- README updates

## [1.3.1] - 2020-03-30

### Changed
- Fixed README and CHANGELOG

## [1.3.0] - 2020-03-30

### Added
- Custom output templates using the templating package

## [1.2.0] - 2020-02-12

### Added
- Config options can now be overridden via annotations.

### Changed
- Migrated from go dep to go modules.
- Migrated the build pipeline from Travis to Github Actions.
- Updated license.
- Upgraded to sensu-plugin-sdk.
- Updated the default icon URL.

### Deprecated
- Environment variables prefixed with `SENSU_` are now deprecated and will be
removed in a future release.

## [1.0.3] - 2019-01-09

### Added
- Use SLACK_WEBHOOK_URL envvar for default value of slack_webhook_url.  Use of envvar by default prevents leaking of sensitive credential into system process table via command argument. This is a backwards compatible change, and the --webhook-url argument can still be used as an override for testing purposes.

### Added
- Adds .bonsai.yml.

## [1.0.2] - 2018-12-04

### Added
- Travis post-deploy script generates a sha512 for packages to be sensu asset compatible. 

## [1.0.1] - 2018-11-30

### Changed
- Corrected binary name in help output

## [1.0.0] - 2018-11-30

### Breaking Changes
- Updated sensu-go version to GA RC SHA.

## [0.1.2] - 2018-11-27

### Changed
- Updated the goreleaser file so that the handler is packaged as a sensu compatible asset.

## [0.1.1] - 2018-11-21

### Breaking Changes
- Updated sensu-go version to beta-8 and fixed some breaking changes that
were introduced (Entity.ID -> Entity.Name).

### Removed
- Removed the vendor directory. Dependencies are still managed with Gopkg.toml.

## [0.0.2] - 2018-11-04

Testing Asset goreleaser pipeline.

## [0.0.1] - 2018-08-17

### Added
- More readme instructions

### Changed
- Repo name `slack_handler` to `slack-handler`

## [0.0.0] - 2018-08-17

### Added
- Slack handler
- /vendor
- goreleaser.yml
- travis.yml
- Gopkg.toml
- Gopkg.lock
- LICENSE
- README.md


main_test.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"os"
	"testing"

	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestFormattedEventAction(t *testing.T) {
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")

	action := formattedEventAction(event)
	assert.Equal("RESOLVED", action)

	event.Check.Status = 1
	action = formattedEventAction(event)
	assert.Equal("ALERT", action)
}

func TestChomp(t *testing.T) {
	assert := assert.New(t)

	trimNewline := chomp("hello\n")
	assert.Equal("hello", trimNewline)

	trimCarriageReturn := chomp("hello\r")
	assert.Equal("hello", trimCarriageReturn)

	trimBoth := chomp("hello\r\n")
	assert.Equal("hello", trimBoth)

	trimLots := chomp("hello\r\n\r\n\r\n")
	assert.Equal("hello", trimLots)
}

func TestEventKey(t *testing.T) {
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")
	eventKey := eventKey(event)
	assert.Equal("entity1/check1", eventKey)
}

func TestEventSummary(t *testing.T) {
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check.Output = "disk is full"

	eventKey := eventSummary(event, 100)
	assert.Equal("entity1/check1:disk is full", eventKey)

	eventKey = eventSummary(event, 5)
	assert.Equal("entity1/check1:disk ...", eventKey)
}

func TestFormattedMessage(t *testing.T) {
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check.Output = "disk is full"
	event.Check.Status = 1
	formattedMsg := formattedMessage(event)
	assert.Equal("ALERT - entity1/check1:disk is full", formattedMsg)
}

func TestMessageColor(t *testing.T) {
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")

	event.Check.Status = 0
	color := messageColor(event)
	assert.Equal("00FF00", color)

	event.Check.Status = 1
	color = messageColor(event)
	assert.Equal("FFFF00", color)

	event.Check.Status = 2
	color = messageColor(event)
	assert.Equal("FF0000", color)
}

func TestMessageStatus(t *testing.T) {
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")

	event.Check.Status = 0
	status := messageStatus(event)
	assert.Equal("Resolved", status)

	event.Check.Status = 1
	status = messageStatus(event)
	assert.Equal("Warning", status)

	event.Check.Status = 2
	status = messageStatus(event)
	assert.Equal("Critical", status)
}

func TestSendMessage(t *testing.T) {
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")

	var apiStub = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := ioutil.ReadAll(r.Body)
		expectedBody := `{"@type":"MessageCard","@context":"https://schema.org/extensions","summary":"Sensu alert card","title":"Sensu - RESOLVED","themeColor":"00FF00","sections":[{"text":"Test","activityTitle":"check1","activitySubtitle":"2021-11-17 02:00","facts":[{"name":"Sender:","value":"Sensu"},{"name":"Status:","value":"Resolved"},{"name":"Entity:","value":"entity1"}]}],"potentialAction":[{"@type":"OpenUri","name":"View in Sensu","targets":[{"os":"default","uri":"http://localhost:3000"}]}]}`
		assert.Equal(expectedBody, string(body))
		w.WriteHeader(http.StatusOK)
		_, err := w.Write([]byte(`1`))
		require.NoError(t, err)
	}))

	config.teamsSender = "Sensu"
	config.teamsSensuURL = "http://localhost:3000"
	config.teamswebHookURL = apiStub.URL
	config.teamsIsTest = "true"
	config.teamsDescriptionTemplate = "{{ .Check.Output }}"
	err := sendMessage(event)
	assert.NoError(err)
}

func TestMain(t *testing.T) {
	assert := assert.New(t)
	file, _ := ioutil.TempFile(os.TempDir(), "sensu-handler-teams-")
	defer func() {
		_ = os.Remove(file.Name())
	}()

	event := corev2.FixtureEvent("entity1", "check1")
	eventJSON, _ := json.Marshal(event)
	_, err := file.WriteString(string(eventJSON))
	require.NoError(t, err)
	require.NoError(t, file.Sync())
	_, err = file.Seek(0, 0)
	require.NoError(t, err)
	os.Stdin = file
	requestReceived := false

	var apiStub = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requestReceived = true
		w.WriteHeader(http.StatusOK)
		_, err := w.Write([]byte(`1`))
		require.NoError(t, err)
	}))

	oldArgs := os.Args
	os.Args = []string{"teams", "-w", apiStub.URL}
	defer func() { os.Args = oldArgs }()

	//Figure out os.Exit(0) issue
	main()
	assert.True(requestReceived)
}


main.go
package main

import (
	"fmt"
	"log"
	"strings"
	"time"

	goteamsnotify "github.com/atc0005/go-teams-notify/v2"
	"github.com/sensu-community/sensu-plugin-sdk/sensu"
	"github.com/sensu-community/sensu-plugin-sdk/templates"
	corev2 "github.com/sensu/sensu-go/api/core/v2"
)

// HandlerConfig contains the Teams handler configuration
type HandlerConfig struct {
	sensu.PluginConfig
	teamswebHookURL          string
	teamsIsTest              string
	teamsSender              string
	teamsSensuURL            string
	teamsDescriptionTemplate string
}

const (
	webHookURL          = "webhook-url"
	isTest              = "is-test"
	sender              = "sender"
	sensuURL            = "sensu-url"
	descriptionTemplate = "description-template"

	defaultIsTest   = "false"
	defaultSensuURL = "http://localhost:3000"
	defaultSender   = "Sensu"
	defaultTemplate = "{{ .Check.Output }}"
)

var (
	config = HandlerConfig{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-teams-handler",
			Short:    "The handler adapted from the Sensu Go Slack handler for notifying a Microsoft Teams channel",
			Keyspace: "sensu.io/plugins/teams/config",
		},
	}

	teamsConfigOptions = []*sensu.PluginConfigOption{
		{
			Path:      webHookURL,
			Env:       "TEAMS_WEBHOOK_URL",
			Argument:  webHookURL,
			Shorthand: "w",
			Secret:    true,
			Usage:     "The webhook url to send messages to",
			Value:     &config.teamswebHookURL,
		},
		{
			Path:      isTest,
			Env:       "TEAMS_IS_TEST",
			Argument:  isTest,
			Shorthand: "t",
			Default:   defaultIsTest,
			Usage:     "Specify if this is a test run",
			Value:     &config.teamsIsTest,
		},
		{
			Path:      sender,
			Env:       "TEAMS_SENDER",
			Argument:  sender,
			Shorthand: "s",
			Default:   defaultSender,
			Usage:     "The name that messages will be sent as",
			Value:     &config.teamsSender,
		},
		{
			Path:      sensuURL,
			Env:       "TEAMS_SENSU_URL",
			Argument:  sensuURL,
			Shorthand: "u",
			Default:   defaultSensuURL,
			Usage:     "A URL to an image to use as the user avatar",
			Value:     &config.teamsSensuURL,
		},
		{
			Path:      descriptionTemplate,
			Env:       "TEAMS_DESCRIPTION_TEMPLATE",
			Argument:  descriptionTemplate,
			Shorthand: "d",
			Default:   defaultTemplate,
			Usage:     "The Teams notification output template, in Golang text/template format",
			Value:     &config.teamsDescriptionTemplate,
		},
	}
)

func main() {
	goHandler := sensu.NewGoHandler(&config.PluginConfig, teamsConfigOptions, checkArgs, sendMessage)
	goHandler.Execute()
}

func checkArgs(_ *corev2.Event) error {
	if len(config.teamswebHookURL) == 0 {
		return fmt.Errorf("--%s or TEAMS_WEBHOOK_URL environment variable is required", webHookURL)
	}

	return nil
}

func formattedEventAction(event *corev2.Event) string {
	switch event.Check.Status {
	case 0:
		return "RESOLVED"
	default:
		return "ALERT"
	}
}

func formattedTitle(event *corev2.Event) string {
	return fmt.Sprintf("%s - %s", config.teamsSender, formattedEventAction(event))
}

func chomp(s string) string {
	return strings.Trim(strings.Trim(strings.Trim(s, "\n"), "\r"), "\r\n")
}

func eventKey(event *corev2.Event) string {
	return fmt.Sprintf("%s/%s", event.Entity.Name, event.Check.Name)
}

func eventSummary(event *corev2.Event, maxLength int) string {
	output := chomp(event.Check.Output)
	if len(event.Check.Output) > maxLength {
		output = output[0:maxLength] + "..."
	}
	return fmt.Sprintf("%s:%s", eventKey(event), output)
}

func formattedMessage(event *corev2.Event) string {
	return fmt.Sprintf("%s - %s", formattedEventAction(event), eventSummary(event, 100))
}

func messageColor(event *corev2.Event) string {
	switch event.Check.Status {
	case 0:
		return "00FF00"
	case 2:
		return "FF0000"
	default:
		return "FFFF00"
	}
}

func messageStatus(event *corev2.Event) string {
	switch event.Check.Status {
	case 0:
		return "Resolved"
	case 2:
		return "Critical"
	default:
		return "Warning"
	}
}

func messageSection(event *corev2.Event) *goteamsnotify.MessageCardSection {

	description, err := templates.EvalTemplate("description", config.teamsDescriptionTemplate, event)
	if err != nil {
		fmt.Printf("%s: Error processing template: %s", config.PluginConfig.Name, err)
	}
	description = strings.Replace(description, `\n`, "\n", -1)

	section := goteamsnotify.NewMessageCardSection()

	section.ActivityTitle = event.Check.Name

	if config.teamsIsTest == "false" {
		section.ActivitySubtitle = time.Now().Local().String()
	} else {
		section.ActivitySubtitle = "2021-11-17 02:00"
	}

	section.AddFactFromKeyValue("Sender:", config.teamsSender)
	section.AddFactFromKeyValue("Status:", messageStatus(event))
	section.AddFactFromKeyValue("Entity:", event.Entity.Name)

	if config.teamsIsTest == "false" {
		section.Text = description
	} else {
		section.Text = "Test"
	}

	return section
}

func messageActionOpenURI(name string, targetURL string) *goteamsnotify.MessageCardPotentialAction {

	action, err := goteamsnotify.NewMessageCardPotentialAction(goteamsnotify.PotentialActionOpenURIType, name)

	if err != nil {
		log.Fatal("error encountered when creating new action:", err)
	}

	action.MessageCardPotentialActionOpenURI.Targets =
		[]goteamsnotify.MessageCardPotentialActionOpenURITarget{
			{
				OS:  "default",
				URI: targetURL,
			},
		}

	return action
}

func sendMessage(event *corev2.Event) error {
	hook := goteamsnotify.NewClient()

	hook.SkipWebhookURLValidationOnSend(true)

	err := hook.Send(config.teamswebHookURL, goteamsnotify.MessageCard{
		Type:             "MessageCard",
		Context:          "https://schema.org/extensions",
		Summary:          "Sensu alert card",
		Title:            formattedTitle(event),
		ThemeColor:       messageColor(event),
		Sections:         []*goteamsnotify.MessageCardSection{messageSection(event)},
		PotentialActions: []*goteamsnotify.MessageCardPotentialAction{messageActionOpenURI("View in Sensu", config.teamsSensuURL)},
	})
	if err != nil {
		return fmt.Errorf("failed to send Teams message: %v", err)
	}

	// FUTURE: send to AH
	fmt.Print("Notification sent to Teams channel\n")

	return nil
}
